fish-rust/src/builtins/pwd.rs

```diff
@@ -0,0 +1,73 @@
+use std::ffi::c_int;
+
+use crate::{
+    builtins::shared::{io_streams_t, BUILTIN_ERR_ARG_COUNT1},
+    ffi::parser_t,
+    wchar::{wstr, WString, L},
+    wchar_ffi::WCharFromFFI,
+    wgetopt::{wgetopter_t, wopt, woption, woption_argument_t},
+    wutil::{wgettext_fmt, wrealpath},
+};
+
+use super::shared::{
+    builtin_print_help, builtin_unknown_option, STATUS_CMD_ERROR, STATUS_CMD_OK,
+    STATUS_INVALID_ARGS,
+};
+
+const short_options: &wstr = L!("LPh");
+const long_options: &[woption] = &[
+    wopt(L!("help"), woption_argument_t::no_argument, 'h'),
+    wopt(L!("logical"), woption_argument_t::no_argument, 'L'),
+    wopt(L!("physical"), woption_argument_t::no_argument, 'P'),
+];
+
+pub fn pwd(parser: &mut parser_t, streams: &mut io_streams_t, argv: &mut [&wstr]) -> Option<c_int> {
+    let cmd = argv[0];
+    let argc = argv.len();
+    let mut resolve_symlinks = false;
+    let mut w = wgetopter_t::new(short_options, long_options, argv);
+    while let Some(opt) = w.wgetopt_long() {
+        match opt {
+            'L' => resolve_symlinks = false,
+            'P' => resolve_symlinks = true,
+            'h' => {
+                builtin_print_help(parser, streams, cmd);
+                return STATUS_CMD_OK;
+            }
+            '?' => {
+                builtin_unknown_option(parser, streams, cmd, argv[w.woptind - 1], false);
+                return STATUS_INVALID_ARGS;
+            }
+            _ => panic!("unexpected retval from wgetopt_long"),
+        }
+    }
+
+    if w.woptind != argc {
+        streams
+            .err
+            .append(wgettext_fmt!(BUILTIN_ERR_ARG_COUNT1, cmd, 0, argc - 1));
+        return STATUS_INVALID_ARGS;
+    }
+
+    let mut pwd = WString::new();
+    let tmp = parser.vars1().get_pwd_slash();
```

Comment:
Previously this was `get(L"PWD")`. This will change the output to add an extra slash. Is there a reason that `parser.vars().get(L"PWD")` couldn't be ported? Maybe we need to add an interface for accessing variables?

Reply:
I think there has been some discussion on this before, but the essence of the issue is that `env_stack_t::get` returns a `maybe_t<env_var_t>` which `autocxx` can't handle. I know `get_pwd_slash` will produce an extraneous trailing slash so I added code to `pop()` the slash. I guess an alternative might be to write some kind of conversion in the C++ codebase from `maybe_t` to something like `unique_ptr`. However, I think these kinds of issues might only be fully resolved once the parser is ported (there are a number of other builtins that rely on parser functionality).

Reply:
You can't rely on popping the slash because PWD might have been set with the trailing slash (even if that's not as correct). Unless I'm misunderstanding.

Reply:
I think you're right. I'm going to push a change to make sure `$PWD` doesn't have a trailing slash.

Reply:
Actually I think it might be more complicated that that. Both `env_stack_t::get` and `env_stack_t::get_pwd_slash` appear to read `$PWD` from `perproc_data.pwd`. `env_stack_t::get_pwd_slash` checks if there is a trailing slash and if there isn't one already, it appends a trailing slash. So we have no way of knowing whether the trailing slash is there "intentionally" or not.

Reply:
Per our discussion, I've introduced a wrapper for `env_stack_t::get` like @ridiculousfish suggested.